ITE/4/README.adoc

```diff
@@ -0,0 +1,510 @@
+    },
+    "byproducts": {
+        "stderr": "",
+        "stdout": "",
+        "return-value": ""
+    },
+    "environment": {
+        "variables": "",
+        "filesystem": "",
+        "workdir": ""
+    }
+}
+```
+
+[[reasoning]]
+== Reasoning
+
+This ITE proposes a change in the URI scheme of artifacts in in-toto metadata.
+The proposed change closely matches the URI structure laid out in RFC 3986. The
+URI structure proposed in that document is widely accepted and is versatile
+enough to allow for a wide variety of references. It is also easy to implement
+and extend support for due to the presence of a large number of standard
+libraries.
+
+[[backwards-compatibility]]
+== Backwards Compatibility
+
+If in-toto metadata is generated using an implementation of in-toto conforming
+to this ITE, verification using a non-conforming implementation can fail. It is
+possible for the verification workflow to progress if the inspections don't
+use generic URIs.
+
+However, a conforming implementation SHOULD be capable of verifying in-toto
+metadata generated using a non-conforming implementation, as an ITE-4 conforming
+system MUST also conform to the actual in-toto specification.
+
+It's also possible two conforming systems may be unable to verify the other's
+in-toto metadata as they're unaware of how to resolve certain URI tokens used by
+the other. This is again because of the possibility of inspections containing
+URIs that the other system is unable to resolve and calculate cryptographic
+information for.
+
+[[security]]
+== Security
+
+As per the in-toto specification, the only direct interaction in-toto tools have
+with artifacts is to record their hashes using one or more cryptographic hash
+algorithms. Other artifact operations such as verifying the artifact rules
+rely on the hashes recorded in the link metadata.
+
+We found that the recording of hashes of abstract resources, some of which may
+live at remote locations, is more complicated than the recording of hashes of
+artifacts in the local filesystem. It is, therefore, important for ITE-4
+compliant in-toto tools to handle the resolving of generic URIs to these
+abstract resources securely. It is quite possible that the contents of these
+resources need to be serialized in some manner that allows for their hashing,
+and these operations when performed unsafely can lead to severe vulnerabilities
+if the resource is controlled by a malicious actor.
+
+Further, it is likely that abstract resources change more frequently, both in
+content and format, and implementors must take care to identify how these
+contents are recorded, as well what specific information is recorded for a
+particular type of entity. Otherwise, the in-toto verification workflow can be
+plagued by failures due to the lack of availability of the artifact as
+previously recorded. For example, implementors who decide how to resolve GitHub
+pull request must decide what information encoded in a pull request must be
```

Comment:
Can you just state that you require such items to be static (non-changing) and declare modifiable items to be out of scope?

Reply:
> Can you just state that you require such items to be static (non-changing) and declare modifiable items to be out of scope?

I may be overthinking here, but do you mean that these resources shouldn't change _in_ the supply chain? Meaning that `MODIFY` doesn't apply to these resources, and any change would require links to be regenerated?

Reply:
I mean that the thing the URI points to should be static, not that you cannot have a MODIFY step or similar.

Reply:
I opened #7 and I've inserted a link in the document to further discuss this, based especially on my findings with implementing this for GitHub.